### PR TITLE
Add disk-monitor support for partition-local stores

### PR DESCRIPTION
**Pull request by lava**
- Add new 'lru_cache::eject()' member function
- Implement disk monitor support for local stores

### :memo: Checklist

- [ ] All user-facing changes have changelog entries.
- [x] The changes are reflected on [docs.tenzir.com/vast](https://docs.tenzir.com/vast), if necessary.
- [x] The PR description contains instructions for the reviewer, if necessary.

### :dart: Review Instructions

Review commits individually.

### CODE REVIEW

File: libvast/src/system/partition.cpp [hunk 1]
```diff
@@ -900,6 +908,22 @@ partition_actor::behavior_type passive_partition(
...
+      if (err)
+        VAST_WARN("{} could not erase partition at", self, self->state.path);
+      vast::ids all_ids;
+      for (const auto& kv : self->state.type_ids) {
+        all_ids |= kv.second;
+      }
+      query q;
+      q.cmd = query::erase{};
+      auto rp = self->make_response_promise<atom::done>();
+      return rp.delegate(self->state.store, q, all_ids);
```
Comment (dominiklohmann) [inline libvast/src/system/partition.cpp:925]: Isn't this just `self->delegate`?

File: libvast/src/system/partition.cpp [hunk 2]
```diff
@@ -515,6 +515,13 @@ active_partition_actor::behavior_type active_partition(
     shutdown<policy::parallel>(self, std::move(indexers));
   });
   return {
+    [self](atom::erase) -> caf::result<atom::done> {
+      // Erase is sent by the disk monitor to erase this partition
+      // from disk, but an active partition does not have any files
+      // on disk, so it should never get selected for deletion.
+      VAST_WARN("{} got erase atom as an active partition", self);
+      return atom::done_v;
```
Comment (dominiklohmann) [inline libvast/src/system/partition.cpp:523]: Should this return an error to the disk monitor instead?

File: libvast/src/system/index.cpp [hunk 3]
```diff
@@ -935,66 +935,73 @@ index(index_actor::stateful_pointer<index_state> self,
...
+                [=](caf::error& err) mutable {
+                  rp.deliver(std::move(err));
+                });
           },
-          [=](caf::error& err) mutable {
-            rp.deliver(std::move(err));
+          [partition_id](const caf::error& err) {
+            // We don't go through the LRU cache to spawn these partitions,
+            // since they are literally the wor auto partition_actor =
+            // self->spawn() self->send(partition_actor, atom::erase_v);
```
Comment (dominiklohmann) [inline libvast/src/system/index.cpp:1001]: This comment looks incomplete.

File: libvast/src/system/local_segment_store.cpp [hunk 4]
```diff
@@ -165,6 +166,18 @@ passive_local_store(store_actor::stateful_pointer<passive_store_state> self,
...
+      VAST_DEBUG("{} erases some ids");
+      if (is_subset(self->state.segment->ids(), xs)) {
+        VAST_VERBOSE("{} gets wholly erased from {}", self, self->state.path);
+        std::error_code err;
+        std::filesystem::remove_all(self->state.path, err);
+        if (err)
+          return caf::make_error(ec::system_error, err.message());
+        // We don't explicitly `quit()` here just in case a query is still in
+        // progress - the mmap will still be valid until unmapped, even if the
+        // underlying file is erased.
```
Comment (dominiklohmann) [inline libvast/src/system/local_segment_store.cpp:178]: I don't understand this comment.

File: libvast/vast/detail/lru_cache.hpp [hunk 5]
```diff
@@ -16,12 +16,16 @@
...
+// The following extensive changes have been applied to the original source:
+//  * Cosmetic refactoring to fit the VAST code style
+//  * Use of exceptions has been removed
+//  * A `factory` member was added to create new entries if a key is
+//    missing from the cache.
+//  * Iteration support
+//  * Member function `exists()` was renamed to `contains()`
+//  * Added member function `resize()`
+//  * Added member function `clear()`
+//  * Added member function `eject()`
```
Comment (dominiklohmann) [inline libvast/vast/detail/lru_cache.hpp:28]: Do these changes need to be reflected in LICENSE.3rdparty?

**Review by dominiklohmann — COMMENTED** [review]
Looks great. I only looked at the code so far, will give this a spin later today. Some early comments for now.

File: libvast/vast/detail/lru_cache.hpp (hunk 5 above)
Comment (lava) [reply libvast/vast/detail/lru_cache.hpp:28]: I don't think so, we're not adding any new third-party code but rather add our own modifications to existing code.

**Review by dominiklohmann — APPROVED** [review]
I've ran this locally for a few hours and things seem to work as expected. Please adjust the old changelog entry before merging.

**Pull request merged**